src/librustc_mir/interpret/terminator.rs

```diff
@@ -182,20 +182,28 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
     }
 
     fn check_argument_compat(
+        rust_abi: bool,
         caller: TyLayout<'tcx>,
         callee: TyLayout<'tcx>,
     ) -> bool {
         if caller.ty == callee.ty {
             // No question
             return true;
         }
+        if !rust_abi {
+            // Don't risk anything
+            return false;
+        }
```

Comment:
This changes the behavior in the case of `(Scalar, Scalar)` pairs, right? (They didn't use to be dependent on the `rust_abi`) -- @eddyb is that what you intended?

Reply:
Yes, scalar pairs are only used in the Rust ABI, two types with the same scalar pair optimization can have incompatible by-value ABI with `extern "C"`.

Reply:
Yes. It also changes behavior on *non-Rust* ABIs for `Scalar`, where we no longer allow any kind of mismatch -- mostly for consistency.